server/src/main/java/org/elasticsearch/common/filesystem/WindowsFileSystemNatives.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.filesystem;
+
+import com.sun.jna.Native;
+import com.sun.jna.WString;
+import com.sun.jna.ptr.IntByReference;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.util.Constants;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.OptionalLong;
+
+/**
+ * {@link FileSystemNatives.Provider} implementation for Windows/Kernel32
+ */
+final class WindowsFileSystemNatives implements FileSystemNatives.Provider {
+
+    private static final Logger logger = LogManager.getLogger(WindowsFileSystemNatives.class);
+
+    private static final WindowsFileSystemNatives INSTANCE = new WindowsFileSystemNatives();
+
+    private WindowsFileSystemNatives() {
+        if (Constants.WINDOWS) {
+            try {
+                Native.register("kernel32");
+                logger.debug("windows/Kernel32 library loaded");
+            } catch (NoClassDefFoundError e) {
+                logger.warn("JNA not found. native methods and handlers will be disabled.");
+            } catch (UnsatisfiedLinkError e) {
+                logger.warn("unable to link Windows/Kernel32 library. native methods and handlers will be disabled.");
+            }
+        }
+    }
```

Comment:
Sorry, forgot to say this earlier.   This constructor makes me a little uneasy.   I wonder if we can assert Windows (rather than using a conditional)? Also, what if register fails/throws? Shouldn't that results in the constructor failing or similar? otherwise the virtual method is invokable, which I suspect could lead to a crash? So, maybe log and allow the link error to propagate out? ( the code will then end up with the no-op provider )

Reply:
I pushed e5a0e2c72ec61ebc46d91ea247d391b74b714dd7, let me know what you think